core/classifiers/CodeClassifier/CodeClassifier.py

```diff
@@ -0,0 +1,663 @@
+# program.
+UNK_TOKEN = 'UNK'
+
+# T is minimum length of a substring that winnowing should match when
+# comparing two programs.
+# Minimum value of T. T is model parameter of winnowing algorithm.
+T_MIN = 3
+# Maximum value of T. T is model parameter of winnowing algorithm.
+T_MAX = 11
+
+# K is length of substring for which hash value is to be generated. Each
+# substring of length K in program will be converted to equivalent hash
+# value (K <= T).
+# Minimum value of K (K <= T). K is model parameter of winnowing algorithm.
+K_MIN = 3
+
+def get_token(program):
+    """Generate tokens for program using tokenize module.
+
+    Args:
+        program: str. Input program for which tokens are to be generated.
+
+    Yields:
+        int. Integer value which represents type of token.
+        str. Token.
+    """
+    for token_id, token_name, _, _, _ in tokenize.generate_tokens(
+            StringIO.StringIO(program).readline):
+        yield (token_id, token_name)
+
+
+def cv_tokenizer(program, var_token=VAR_TOKEN):
+    """Custom tokenizer of Python program for CountVectorizer in sci-kit.
+
+    Args:
+        program: str. Input program for which tokens are to be generated.
+        var_token: str. Replace variable and method names with this string when
+            tokenizing data.
+
+    Returns:
+        list(str). A list containing tokens of input program.
+    """
+    token_program = []
+    for token_id, token_name in get_token(program):
+        # Ignore all newline, comment and empty string tokens.
+        if (token_id == token.N_TOKENS or token_id == 54
+                or token_name.strip() == ''):
+            continue
+        elif token_id == token.NAME:
+            # If token_id is tokens.NAME then only add if it is a python
+            # keyword. Treat all variables and methods same.
+            if token_name in keyword.kwlist:
+                token_program.append(token_name)
+            else:
+                token_program.append(var_token)
+        else:
+            token_program.append(token_name)
+
+    return token_program
+
+
+def generate_token_to_id(
+        data, threshold, var_token=VAR_TOKEN, unk_token=UNK_TOKEN):
+    """Generates a list of valid tokens and assigns a unique ID to each
+    token.
+
+    Args:
+        data: dict. A dictionary containing training_data.
+        threshold: int. Ignore a token if it appears less than 'threshold' times
+            in dataset.
+        var_token: str. Replace variable and method names with this string when
+            tokenizing data.
+        unk_token: ste. Replace unknown tokens (ignored tokens) with this string
+            when tokenizing the data.
+
+    Returns:
+        dict. A dictionary containing tokens mapped to a unique ID.
+    """
+    # All unique tokens and number of time they occur in dataset. A token
+    # can be a keyword, an identifier, a constant, an operator.
+    alphabet = {}
+
+    for pid in data:
```

Comment:
What kind of specific? I mean it is training data after all. It has source program and other things which are used for feature extraction by winnowing or svm.